src/scanner.cc

```diff
@@ -108,6 +111,94 @@ struct Scanner {
     }
   }
 
+  bool is_escapable_sequence(TSLexer *lexer) {
+    // Note: remember to also update the escape_sequence rule in the
+    // main grammar whenever changing this method
+    auto letter = lexer->lookahead;
+
+    if (letter == 'n' ||
+        letter == 'r' ||
+        letter == 't' ||
+        letter == 'v' ||
+        letter == 'e' ||
+        letter == 'f' ||
+        letter == '\\' ||
+        letter == '$' ||
+        letter == '"') {
+      return true;
+    }
+
+    // Hex
+    if (letter == 'x') {
+      advance(lexer);
+      return isxdigit(lexer->lookahead);
+    }
+
+    // Unicode
+    if (letter == 'u') {
+      advance(lexer);
+      return lexer->lookahead == '{';
+    }
+
+    // Octal
+    return isdigit(lexer->lookahead) && lexer->lookahead >= 0 && lexer->lookahead <= 7;
+  }
+
+  bool scan_encapsed_part_string(TSLexer *lexer, bool is_after_variable) {
+    lexer->result_symbol = ENCAPSED_STRING_CHARS;
+
+    for (bool has_content = false;; has_content = true) {
+      lexer->mark_end(lexer);
+
+      switch (lexer->lookahead) {
+        case '"':
+          return has_content;
+        case '\0':
+          return false;
+        case '\\':
+          advance(lexer);
+          if (is_escapable_sequence(lexer)) {
+            return has_content;
+          }
+          break;
+        case '$':
+          advance(lexer);
+          if (isalpha(lexer->lookahead) || lexer->lookahead == '_') {
```

Comment:
Should this handle the case of variable variables (`$$a`)?

Perhaps it's useful to split `is_alpha(a) || a == '_'` to a separate function.

Reply:
See also [`iswalpha`](https://www.cplusplus.com/reference/cwctype/iswalpha/) in `<cwctype>`

Reply:
Good point! I've extended the detection to support arbitrarily many sequential dollar signs.
Also, I've made a separate function to check for valid names in PHP: `is_valid_name_char`